lib/active_record_host_pool/connection_handling.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module ActiveRecordHostPool
+  # ActiveRecord 6.0 introduces multiple database support. This introduces a
+  # change where dirtying operations on a Mysql2Adapter instance calls out to
+  # ActiveRecord::Base.clear_query_caches_for_current_thread. This will
+  # for any active connections whose query caches need to be cleared. This
+  # involves iterating over every connection handler's connection pool
+  # looking for active active connections.
+  #
+  # This exposes an issue with active active_record_host_pool since by
+  # referencing a connection it causes the underlying Mysql2Adapter instance
+  # to potentially change its _host_pool_current_database. This means
```

Comment:
How does the `_host_pool_current_database` change during a call to `#clear_query_caches_for_current_thread`?

Reply:
That is a great question. Let me see if I can clearly and concisely communicate this. 

ARHP pools connections together by host essentially (the pool key also includes port/socket/slave/user but those are unimportant right now). This means that a connection pool for two databases might look like this:

```ruby
host1:
- :database:
    :name: primary
    :connection: (ARHP::ConnectionProxy1)
      :connection: connectionA
- :database:
    :name: shard1
    :connection: (ARHP::ConnectionProxy2)
       :connection: connectionA
```

Now let's say you have two models:

```ruby
class ShardedModel < ActiveRecord::Base
    # uses shard1 database
end

class UnshardedModel < ActiveRecord::Base
   # uses primary1 database
   not_sharded 
end 
```

When you reference any method on `ShardedModel.connection` the code path goes through ARHP's connection proxy, and this changes `@_host_pool_current_database` to point to the "primary" database. This will make the next query that gets executed switch databases. And when you reference code that goes through `UnshardedModel.connection` it does the same thing except changes `@_host_pool_current_database` to point to the "shard1" database.

In Rails 6, they [updated methods like `insert` on the connection adapter](https://github.com/rails/rails/blob/b9d00b37b30840bc95421b6c5fad4a4ea3c71eae/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb#L21 
) to call out to:

```
ActiveRecord::Base.clear_query_caches_for_current_thread if @query_cache_enabled
```

This will end up looking for active connections for the current thread so its cache can be cleared.

Internally, ActiveRecord only keeps reference to one active connection at a time for the current thread. Because ARHP implements a ConnectionProxy around a shared Mysql2Adapter instance this is where the problem comes in. The active connection is the same for both the `ShardedModel` and the `UnshardedModel`. However, what is stored is the ARHP::ConnectionProxy instance. So if ARHP::ConnectionProxy1 is stored as the current thread's active connection then when happens is:

```ruby
UnshardedModel.create!(name: "foo")
# -> this references the connection internally changes the database name on shared Mysql2Adapter instance. It now points to "shard1"
# --> the adapter gets insert called on it
# --->  insert calls ActiveRecord::Base.clear_query_caches_for_current_thread
# -----> find active connection for thread. 
# -----> returns ARHP::Connection1
# -----> call clear_cache on ARHP::Connection1
# ---------> the this reference to the connection internally changes the database name on the shared Mysql2Adapter. It now points to "primary"
# ---> execute INSERT SQL # BLOWS UP, database doesn't know about this table
```
 
Does this help?

Reply:
It sure does, thank you 👍🏼